### PR TITLE
internal/{v4,v5}: implement read-only mode

### PR DESCRIPTION
**Pull request by rogpeppe**
We'd like to be able to stop people making changes
to the charm store while we're upgrading, so
add a read-only mode to allow that possibility.

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2288/

File: internal/router/util.go [hunk 1]
```diff
@@ -82,7 +82,9 @@ func errorToResp1(err error) (int, interface{}) {
 		status = http.StatusNotFound
 	case params.ErrBadRequest, params.ErrInvalidEntity:
 		status = http.StatusBadRequest
-	case params.ErrForbidden, params.ErrEntityIdNotAllowed:
+	case params.ErrForbidden,
+		params.ErrEntityIdNotAllowed,
+		params.ErrReadOnly:
 		status = http.StatusForbidden
```
Comment (mhilton) [inline internal/router/util.go:88]: I feel this is more of a 503 Service Unavailable type situation. It should almost certainly be a 5XX error. There's nothing about the request that means it's forbidden.

**Review by mhilton — CHANGES_REQUESTED** [review]
This doesn't look sufficient to me. It doesn't cover stats updates , for instance. and I suspect there are other unexpected places we write to the database.

File: internal/router/util.go (hunk 1 above)
Comment (rogpeppe) [reply internal/router/util.go:88]: You're forbidden from writing to the store. It seems that StatusForbidden might be appropriate. 503 doesn't seem quite right to me, as much of the service is actually available.

**Review by mhilton — APPROVED** [review]
As discussed

**Comment by rogpeppe** [comment]
$$merge$$
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/charmstore-merge

**Pull request merged**